ocf_data_sampler/numpy_sample/collate.py

```diff
@@ -0,0 +1,79 @@
+from ocf_data_sampler.numpy_sample import NWPSampleKey
+
+import numpy as np
+import logging
+from typing import Union
+
+logger = logging.getLogger(__name__)
+
+
+def stack_np_examples_into_sample(dict_list):
```

Comment:
Suggestion for refactoring original function - just for any modification / maintainability concerns considering dealing with overall processing stages for nested data. Worth noting that core behaviour and related testing remains exactly the same.

```python

def stack_np_examples_into_sample(dict_list):
    if not dict_list:
        raise ValueError("Input empty")

   # Extract keys from first dict - structure
    sample = {}
    sample_keys = list(dict_list[0].keys())

   # Process - handle NWP separately due to nested structure
    for sample_key in sample_keys:
        if sample_key == "nwp":
            sample["nwp"] = process_nwp_data(dict_list)
        else:
            sample[sample_key] = stack_data_for_key(dict_list, sample_key)
    return sample


def process_nwp_data(dict_list):
    nwp_sample = {}
    nwp_sources = dict_list[0]["nwp"].keys()

   # Stack data for each NWP source independently
    for nwp_source in nwp_sources:
        nwp_sample[nwp_source] = stack_nested_keys(dict_list, "nwp", nwp_source)
    return nwp_sample


def stack_data_for_key(dict_list, sample_key):
    # Stack arrays for given key across all dicts
    return stack_data_list([d[sample_key] for d in dict_list], sample_key)


def stack_nested_keys(dict_list, parent_key, child_key):
    # Processes two-level nested data 
    # Preserves hierarchy whilst stacking arrays
    nested_keys = dict_list[0][parent_key][child_key].keys()
    return {
        key: stack_data_list(
            [d[parent_key][child_key][key] for d in dict_list],
            key,
        )
        for key in nested_keys
    }

```